src/cls/rbd/cls_rbd.cc

```diff
@@ -2652,9 +2652,9 @@ int metadata_list(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
   map<string, bufferlist> data;
   string last_read = metadata_key_for_name(start_after);
   int max_read = max_return ? MIN(RBD_MAX_KEYS_READ, max_return) : RBD_MAX_KEYS_READ;
-  bool more;
+  bool more =  true;
 
-  do {
+  while (more && data.size() < max_return) {
```

Comment:
i think we should put
```c++
do {
 ...
} while (more && max_read);
```

because, i think, a `0` `max_return` instructs `metadata_list()` to use the max possible limits, not asking it to return right away. see https://github.com/ceph/ceph/pull/17499/files#diff-e393a69c5ca0d5ea740c0d60d68e55f5R2654.

Reply:
Negative -- 0 implies ```RBD_MAX_KEYS_READ```

Reply:
that's what i mean by "the max possible limits".

Reply:
... sorry, cleaned it up some more and re-pushed